Automatically detect the user’s time zone

kuv2707 opened the pull request:
This PR adds support for the web app to automatically detect the user's time zone using 
`new Intl.DateTimeFormat().resolvedOptions().timeZone` and offer to update the user profile time zone with it if it is found to be inconsistent with the browser time zone, and if the option to do so is enabled.

- If the user manually chooses an inconsistent time zone, the option to offer time zone update is disabled.
- If the user manually chooses a consistent time zone, the option is enabled.
- There is a check box to enable/disable this option apart from the automatic behavior. 
  If the option is enabled via the checkbox and the time zone is found inconsistent, a navbar alert is shown on refreshing.

Fixes: #16957 


<details>
<summary> Screenshots </summary>

<img width="1470" alt="image" src="https://github.com/user-attachments/assets/0e28c893-8ce0-43d8-8d7e-2dd1c455be1e" />

<img width="1032" alt="image" src="https://github.com/user-attachments/assets/9e597716-750d-47ea-9803-cb0b9812067d" />

<img width="435" alt="image" src="https://github.com/user-attachments/assets/d0d47518-58e8-44f6-82d7-ba686d339d0e" />

<img width="802" alt="image" src="https://github.com/user-attachments/assets/9ba34bd0-dbc7-4585-b6b4-5f040b53e484" />




</details>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

kuv2707 commented:
@zulipbot add "buddy review"

@pratik-pc Please have a look!

pratik-pc commented:
[Screencast from 17-06-24 03:33:59 AM IST.webm](https://github.com/zulip/zulip/assets/33497322/aeebed1b-9462-481d-b4c1-0c7ec27a1312)

@kuv2707 This isn't working for me

kuv2707 commented:
@pratik-pc It is not supposed to work that way. You can read the issue description for the working.

kuv2707 commented on web/src/navbar_alerts.ts:
```diff
@@ -188,6 +200,13 @@ export function initialize(): void {
             data_process: "notifications",
             rendered_alert_content_html: render_desktop_notifications_alert_content(),
         });
+    } else if (should_offer_to_update_timezone()) {
+        open({
+            data_process: "time_zone_update_offer",
+            rendered_alert_content_html: render_time_zone_update_offer_content({
+                browser_time_zone,
+            }),
+        });
```
I am unsure of the priority of this alert, so I've put it here for now.

timabbott replied:
Probably above `server_needs_upgrade` would be good?

pratik-pc commented:
[Screencast from 17-06-24 02:16:33 PM IST.webm](https://github.com/zulip/zulip/assets/33497322/c98b38e5-6225-46ac-98d6-1aaa9965fd8c)

The spacing increases after selecting an option from the banner.

Also the banner seems to appear in anonymous login.

kuv2707 commented:
@pratik-pc Thanks for pointing these out!

- Fixed the spacing glitch by triggering `resize` on window, like other alerts do.
- Only showing the alert if `page_params.is_spectator` is false.

pratik-pc commented:
This looks good to me now!
Works in manual testing.
For further review stage, tested in anonymous login, narrow window, both banner button, timezone change reflects correctly on banner confirmation, changing system timezone, reset of checkbox on changing timezone.

pratik-pc commented:
@zulipbot remove "buddy review"

pratik-pc commented:
@zulipbot add "mentor review"

kuv2707 commented:
I think the failing test is not relevant to the PR.

```
gpg: directory '/tmp/tmp.ZfUwfdxKzj/.gnupg' created
gpg: keybox '/tmp/tmp.ZfUwfdxKzj/.gnupg/pubring.kbx' created
gpg: /tmp/tmp.ZfUwfdxKzj/.gnupg/trustdb.gpg: trustdb created
gpg: key 624CF77434839225: public key "Groonga Key (Groonga Official Signing Key) <packages@groonga.org>" imported
gpg: Total number processed: 1
gpg:               imported: 1
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:02 --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:03 --:--:--     0curl: (7) Couldn't connect to server

Zulip installation failed (exit code 7)!
```

timabbott commented on zerver/models/users.py:
```diff
@@ -72,6 +72,7 @@ class UserBaseSettings(models.Model):
     display_emoji_reaction_users = models.BooleanField(default=True)
     twenty_four_hour_time = models.BooleanField(default=False)
     starred_message_counts = models.BooleanField(default=True)
+    automatically_detect_time_zone = models.BooleanField(default=True)
```
Please start an "api design" conversation for what to call this. I suspect at the very least, this should have `web_` as a prefix in its name. Maybe `web_suggest_update_timezone` would be more precise?

kuv2707 replied:
I've initiated a [conversation](https://chat.zulip.org/#narrow/stream/378-api-design/topic/Automatically.20detect.20user's.20timezone) there. I'll change it to `web_suggest_update_timezone` for the time being

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -11104,6 +11104,12 @@ paths:
+                automatically_detect_time_zone:
+                  description: |
+                    Option to detect browser time zone and offer to update the user
+                    setting time zone with it.
```
Please read our "new application feature" tutorial section on API changes and documenting this; or use `git log` on this file to see a lot of examples of doing it properly. You need a `**Changes**` entry, etc. And it needs to be very carefully written prose aligning with existing patterns.

kuv2707 replied:
Documentation is new territory for me and I had planned to address it once the core functionality is in a good state. Thanks for the suggestion, I'll change it

kuv2707 replied:
I have updated the documentation and the changelog

timabbott commented on zerver/tests/test_audit_log.py:
```diff
@@ -1178,7 +1178,7 @@ def test_user_group_creation(self) -> None:
             event_time__gte=now,
             event_type=RealmAuditLog.USER_GROUP_DIRECT_USER_MEMBERSHIP_ADDED,
-        )
+        ).order_by("id")
```
Why is this changed here?

kuv2707 replied:
I don't understand why this entirely unrelated test was failing due to the changes, but looking at this test, we expect the log entries to be sorted, and without using `order_by`, there's no guarantee that the result would be sorted as we expect. 

I guess this would have come up sooner or later.

kuv2707 replied:
I have removed it since I am unable to reproduce it. I wonder what caused it in the first place.

timabbott commented on web/src/timerender.ts:
```diff
@@ -17,7 +17,9 @@ import {user_settings} from "./user_settings";
-export let display_time_zone = new Intl.DateTimeFormat().resolvedOptions().timeZone;
+export let display_time_zone = new Intl.DateTimeFormat()
+    .resolvedOptions()
+    .timeZone.replace("Calcutta", "Kolkata");
```
How did you find this one? Do we have any reason to expect this is the only mismatch of this form?

kuv2707 replied:
I live in this timezone, therefore noticed the bug here. I looked it up and it seems like this is a known issue with this renaming (and couldn't find others).
Our backend correctly expects the new name `Kolkata`, but Chrome stuck to the old British-era `Calcutta`.

It would make more sense to use the hour offset instead of the name, which we currently don't do in the project.

kuv2707 replied:
I think this warrants a CZO discussion regarding the implementation.

kuv2707 replied:
Update: Comparing offsets is a perfectly fine way, and there are no issues with DST.

kuv2707 commented:
@zulipbot remove "has conflicts"

alya commented:
@timabbott adding "integration review" too, since you have a prior review and there are replies to your feedback.

kuv2707 commented:
The merge conflict is regarding the feature level, but I am not sure at what level to put this now.

prakhar1144 commented:
> The conflict is regarding the feature level, but I am not sure at what level to put this now.

Feature level 280

kuv2707 commented:
@timabbott Please have a look

kuv2707 commented:
@zulipbot remove "mentor review"

gnprice commented on web/src/navbar_alerts.ts:
```diff
@@ -147,8 +151,31 @@ export function get_demo_organization_deadline_days_remaining(): number {
+    if (uncanonicalized_target_timezone === "") {
+        return false;
+    }
+    const canonocalized_browser_timezone = new Intl.DateTimeFormat("en-US", {
+        timeZone: timerender.display_time_zone,
+    }).resolvedOptions().timeZone;
+    const canonicalized_target_timezone = new Intl.DateTimeFormat("en-US", {
+        timeZone: uncanonicalized_target_timezone,
+    }).resolvedOptions().timeZone;
+    return canonocalized_browser_timezone === canonicalized_target_timezone;
```
Cool, this logic looks right.

This should get tests. One test case should use two different names for the same time zone, and confirm that they compare as equal — for example Asia/Kolkata and Asia/Calcutta, or Europe/Kyiv and Europe/Kiev.

Another test should confirm we don't have the bug that was discussed [in the chat thread](https://chat.zulip.org/#narrow/stream/378-api-design/topic/Automatically.20detect.20user's.20timezone/near/1921455). To do that, take the three time zones America/Los_Angeles, America/Phoenix, and America/Denver, and confirm that America/Phoenix is different from both of the others. For good measure and to help make the point of the test clear to someone reading the test, confirm that the current time in America/Phoenix *does* equal the current time in one of those two other time zones.

kuv2707 replied:
I have added some tests.

andersk commented on web/src/navbar_alerts.ts:
```diff
@@ -147,8 +151,31 @@ export function get_demo_organization_deadline_days_remaining(): number {
+    if (uncanonicalized_target_timezone === "") {
+        return false;
+    }
+    const canonocalized_browser_timezone = new Intl.DateTimeFormat("en-US", {
```
```suggestion
    const canonicalized_browser_timezone = new Intl.DateTimeFormat("en-US", {
```

kuv2707 replied:
Oops, fixed

andersk commented on web/src/navbar_alerts.ts:
```diff
@@ -147,8 +151,31 @@ export function get_demo_organization_deadline_days_remaining(): number {
+
+export function should_offer_to_update_timezone(): boolean {
+    // This offer is only for logged-in users with the setting enabled.
+    return Boolean(
```
This is already a boolean and does not need to be converted.

```suggestion
    return (
```

kuv2707 replied:
Somehow eslint wasn't happy about it, so I had to put a `Boolean` there

![image](https://github.com/user-attachments/assets/dff3b079-eda0-4bae-8b21-18e8fc463474)

andersk replied:
Do you see this error from `tools/lint -g frontend`? If not, your ESLint extension is probably stuck in an inconsistent state. Try **ESLint: Restart ESLint Server** in the command palette.

kuv2707 replied:
I don't see it there.
I'll remove the `Boolean`

andersk commented on web/src/navbar_alerts.ts:
```diff
@@ -147,8 +151,31 @@ export function get_demo_organization_deadline_days_remaining(): number {
+        return false;
+    }
+    const canonocalized_browser_timezone = new Intl.DateTimeFormat("en-US", {
+        timeZone: timerender.display_time_zone,
```
This wil break if, in the future, we allow `timerender.display_time_zone` to be updated based on the user setting (which is the plan after this: #16958). We may actually want to avoid passing a `timeZone` here, or make a separate `timerender.browser_time_zone` variable from which `timerender.display_time_zone` is initialized.

(Also, `timerender.display_time_zone` has already been canonicalized.)

kuv2707 replied:
I have exported `browser_time_zone` from `timerender.ts`

andersk commented on web/src/settings_account.js:
```diff
@@ -774,9 +775,28 @@ export function set_up() {
         e.stopPropagation();
+        const chosen_timezone = this.value;
+        const data = {timezone: chosen_timezone};
+        if (chosen_timezone !== timerender.display_time_zone) {
```
We need to canonicalize `chozen_timezone` for this test. (It may have been server-canonicalized but it is not browser-canonicalized.)

The `if` can be simplified away:

```js
data.web_suggest_update_timezone = chosen_timezone === timerender.display_time_zone;
```

kuv2707 replied:
Done

kuv2707 commented:
@gnprice @andersk Thanks for the feedback. I have addressed it.

Please have a look.

gnprice commented on web/tests/timerender.test.js:
```diff
@@ -642,3 +642,44 @@ run_test("should_display_profile_incomplete_alert", () => {
+
+    const dst_date = new Date("Sat, 17 Jul 2024 11:05:12 GMT");
+    assert.equal(
+        get_offset_difference_at_date("America/Los_Angeles", "America/Phoenix", dst_date),
```
Another angle on this point: the current version of these tests would continue to pass even if `is_browser_timezone_same_as` always returned true, and even if `browser_canonicalize_timezone` always returned `""`.

So we should have a test case that demonstrates this can tell time zones apart, as well as tell when they're the same.

And then to make the test as useful as possible, it's good to pick a case where it would have been relatively easy for a bug to cause us to not tell the time zones apart. In particular the potential bug discussed in that chat thread is a good one to test for.

kuv2707 replied:
I have added tests which check whether the canonicalized forms of "America/Denver", "America/Phoenix" and "America/Los_Angeles" are not equal. Since at any time in a year, two of them will have the same offset, so the same tests also assert that timezones are treated different even if they have the same offset. 

- I have also added a test which asserts that the current time in Phoenix matches one of the timezones in the other two.
- I have picked the case of Phoenix, Denver and LA as discussed in the chat thread.

A trivial assumption is that `browser_time_zone` is never an empty string.

gnprice replied:
Thanks for the revision! These `timerender` tests all look good to me.


> A trivial assumption is that `browser_time_zone` is never an empty string.

I don't understand this remark. Are you saying that's an assumption that your code is making? What would happen if that assumption were to turn out not to always hold?

kuv2707 replied:
The reason for this remark is that we return an empty string from `browser_canonicalize_timezone` for invalid timezones (which includes empty string), so this test would pass in that case if `browser_time_zone()` is `""`
```
assert.equal(timerender.is_browser_timezone_same_as(timerender.browser_time_zone()), true);
```

which is technically fine.

kuv2707 replied:
> What would happen if that assumption were to turn out not to always hold?
This would reflect deeper problems with the browser, since `new Intl.DateTimeFormat().resolvedOptions().timeZone` will not return `""` (it always returns a valid timezone string)

kuv2707 commented:
@timabbott Please have a look

timabbott commented on web/src/timerender.ts:
```diff
@@ -18,7 +18,8 @@ import * as util from "./util.ts";
 let next_timerender_id = 0;
 
-export let display_time_zone = new Intl.DateTimeFormat().resolvedOptions().timeZone;
+export const browser_time_zone = new Intl.DateTimeFormat().resolvedOptions().timeZone;
```
I'm not sure it's correct to define this as `const`; we might want to have `is_browser_timezone_same_as` recompute it later. I don't know exactly how this works, but I think if you change your timezone in your browser's settings, it doesn't reload existing tabs, but this will return a new value?

kuv2707 replied:
Changing the timezone indeed doesn't reload tabs, and changes in the timezone are dynamically reflected. 

We can make `browser_time_zone` a function that always calculates the current timezone.

timabbott commented on web/src/navbar_alerts.ts:
```diff
@@ -251,6 +272,26 @@ export function initialize(): void {
+            url: "/json/settings",
+            data: {timezone: browser_time_zone},
+        });
+        $(window).trigger("resize");
```
Why does this trigger a resize? It also seems a bit odd that we don't check if the API request succeeded; we probably do want to support an error of some sort. (Probably including only hiding the alert if the request succeeded, in a `success` handler).

kuv2707 replied:
It appears that in the existing code, any change to the navbar alert is followed by a resize. If a resize is not triggered, it leaves the page in a glitchy state.

I have changed it to only hide the alert if the request succeeds.

kuv2707 replied:
And for the error, maybe we can show a toast of some sort? In fact, we can show a toast for both success and error.

kuv2707 replied:
I have started a [discussion](https://chat.zulip.org/#narrow/channel/101-design/topic/Showing.20a.20feedback.20widget.20after.20a.20prompt.20to.20update.20timezone) regarding how to handle the errors.

timabbott commented on web/src/settings_account.ts:
```diff
@@ -832,8 +833,15 @@ export function set_up(): void {
+        $("#automatically_offer_update_time_zone").prop(
+            "checked",
+            data.web_suggest_update_timezone,
+        );
```
Do we need to check it here? I'd expect that the `server_events_dispatch` handler for the server executing the settings change would update the UI. If not, that's a bug we need to investigate.

kuv2707 replied:
Sure, I've shifted it to `server_events_dispatch`.

timabbott commented:
I think I've flagged all the implementation concerns I have above. I expect once the above feedback is resolved, I'll do a final round of manual testing and then hopefully integrate this. Feel free to ping me in DMs; I'm trying to do database changes fairly promptly to manage renumbering effort.

alya commented:
I like it!

It should be "time zone", not "timezone" in the feedback widget.

alya commented:
For the error, can we make it specific?

alya commented:
Let's make the confirmation:

## Time zone updated
Your time zone was updated to {time zone}.

alya commented on web/templates/navbar_alerts/time_zone_update_offer.hbs:
```diff
@@ -0,0 +1,10 @@
+<div data-step="1">
+    {{t "Your computer's time zone has changed. Update your profile's time zone to {browser_time_zone}?" }}
```
I think this can just be "your time zone".

timabbott replied:
Yeah, "your profile time zone" could be OK, but probably your briefer version is better.

kuv2707 replied:
Changed to "your time zone"

kuv2707 replied:
Changed to "your time zone"

kuv2707 commented:
Following the recent discussions, I have:
- Addressed the feedback left here.
- Changed the setting label to "Disable prompts when computer's time zone is different".
- Added a button in the success toast to open profile settings ([suggested here](https://chat.zulip.org/#narrow/channel/101-design/topic/Showing.20a.20feedback.20widget.20after.20a.20prompt.20to.20update.20timezone/near/1998290)). It seemed reasonable UX wise.
- Documented the new setting.


Regarding making the error message more specific, I think the only probable error here would be when the web app can't connect to the server, and we already have a pretty loud error mechanism for that. 
The error messages in the toast follow the pattern of the rest of the codebase (where the messages are not very precise).

I have updated the screenshots in the PR description.

@timabbott Please have a look.

timabbott commented on web/src/server_events_dispatch.js:
```diff
@@ -793,6 +793,7 @@ export function dispatch_normal_event(event) {
                 "starred_message_counts",
                 "web_navigate_to_sent_message",
                 "enter_sends",
+                "web_suggest_update_timezone",
```
A good follow-up PR would be to sort these alphabetically.

timabbott commented on web/src/navbar_alerts.ts:
```diff
@@ -11,18 +11,25 @@ import render_insecure_desktop_app_alert_content from "../templates/navbar_alert
 import * as unread from "./unread.ts";
 import * as unread_ops from "./unread_ops.ts";
 import * as unread_ui from "./unread_ui.ts";
+import * as user_settings from "./user_settings.ts";
```
This import is typically `import {user_settings} from "./user_settings.ts"` to avoid `user_settings.user_settings.` nonsense.

alya commented:
The tooltip needs a delay. I also wonder if we should point it downwards, to avoid hiding the time zone.

alya commented:
Really, our standard approach would not involve a tooltip at all -- just a `?` link to the help center.

alya commented:
The profile settings button on the confirmation toast. seems unnecessary. Let's drop it.

The confirmation is also missing a `.` at the end.

timabbott commented on help/change-your-timezone.md:
```diff
@@ -16,4 +20,7 @@ You can change the time zone Zulip displays to others.
 2. Under **Profile**, select a time zone.
 
+To stop receiving prompts when your computer’s time zone differs from the
+device’s, check the box for '**Disable prompts when the computer’s time zone is different**'.
```
I dropped the /help/ commit, since this isn't our standard style for doing /help/ documentation. @laurynmm can you put up a PR for it?

timabbott commented on web/src/tippyjs.ts:
```diff
@@ -812,4 +812,16 @@ export function initialize(): void {
+        onHidden(instance) {
+            instance.destroy();
+        },
+    });
```
I dropped the tooltip from the PR, since I think we want a slightly different implementation, and can just add that in a follow-up PR.

timabbott commented:
I made the changes suggested above and am merging this; we'll probably want to do those fixups over the next week, but they seem small and it'll be easier to review them in isolation without having to re-check the main work in this PR.

Huge thanks for all the work on this @kuv2707 and everyone else who helped review it!

kuv2707 commented:
@timabbott  A failing puppeteer test prevented the merge. It didn't seem to to be related to the changes here, and I can't reproduce it anymore.

timabbott commented:
Rebased and merged, thanks @kuv2707! (Next time, feel free to DM for this sort of thing, as it can help save extra renumbers).
